sdk/core/azure-core/src/main/java/com/azure/core/util/CoreUtils.java

```diff
@@ -376,4 +386,89 @@ public static Context mergeContexts(Context into, Context from) {
+            BinaryDataContent content = BinaryDataHelper.getContent(responseBody);
+            if (content instanceof FluxByteBufferContent) {
+                content.toFluxByteBuffer()
+                    .ignoreElements()
+                    .block();
+            } else if (content instanceof InputStreamContent) {
+                try (InputStream stream = content.toStream()) {
+                    stream.skip(Long.MAX_VALUE);
+                } catch (IOException e) {
+                    throw LOGGER.logExceptionAsError(new UncheckedIOException(e));
+                }
+            }
+            // Other possible types are already read into memory.
+        }
+    }
+
+    /**
+     * This duplicates buffers for the retry policy. From storage:
+     * Clone the original request to ensure that each try starts with the original (unmutated) request. We cannot
+     * simply call httpRequest.buffer() because although the body will start emitting from the beginning of the
+     * stream, the buffers that were emitted will have already been consumed (their position set to their limit),
+     * so it is not a true reset. By adding the map function, we ensure that anything which consumes the
+     * ByteBuffers downstream will only actually consume a duplicate so the original is preserved. This only
+     * duplicates the ByteBuffer object, not the underlying data.
+     * @param httpRequest The http request
+     */
+    // TODO (kasobol-msft) this is from storage retry policy. What's best way to do this in core ?
+    public static void duplicateBuffers(HttpRequest httpRequest) {
+        /*
+        This is from storage:
+        Flux<ByteBuffer> bufferedBody = (httpRequest.getBody() == null)
+            ? null
+            : httpRequest.getBody().map(ByteBuffer::duplicate);
+        httpRequest.setBody(bufferedBody); */
+
+        BinaryData content = httpRequest.getContent();
+        if (content != null) {
+            BinaryDataContent binaryDataContent = BinaryDataHelper.getContent(content);
+            if (binaryDataContent instanceof FluxByteBufferContent) {
+                Flux<ByteBuffer> bufferedBody = binaryDataContent.toFluxByteBuffer().map(ByteBuffer::duplicate);
+                httpRequest.setBody(bufferedBody);
+            } else if (binaryDataContent instanceof InputStreamContent) {
+                InputStream inputStream = binaryDataContent.toStream();
+                if (inputStream.markSupported()) { // TODO (kasobol-msft) what do we do otherwise ?
+                    try {
+                        inputStream.reset();
+                    } catch (IOException e) {
+                        throw LOGGER.logExceptionAsError(new UncheckedIOException(e));
+                    }
+                }
+            }
+                // TODO (kasobol-msft) in memory types should be fine, what to do with stream and file?
+                // Perhaps BinaryData.copy/HttpRequest.copy should take care of this?
```

Comment:
Is this something we should tackle? On one side RetryPolicy should be able to retry things. However we could be passed streams that may not be replayable. Should we assume them to be, read them in memory and make them replayable (which would potentially take the customer by surprise)?

Reply:
Yes, I think so, probably best would be to just assert that stream is seekable and let user implement this on their side - they may have better ideas than buffering. FWIW .NET SDK requires streams to be seekable.

@alzimmermsft @rickle-msft this questions reminds me that we used to require seekable streams (in a bit convoluted way) [here](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/Utility.java#L218-L224) . But since then we made an effort to migrate Storage to use [Core's converter](https://github.com/Azure/azure-sdk-for-java/blob/457609bdddb67507ee4ac652336af0a2ff94a344/sdk/core/azure-core/src/main/java/com/azure/core/util/FluxUtil.java#L202) which doesn't seem to involve any mark/reset logic. Could you guys double check that?